### ISSUE TITLE
How to pack all  html component as one file?

### ISSUE THREAD

**Issue by blueforest** [ractivejs/component-spec#9]
My component like this:

```
<p></p>
<script>
componet.experts= {
}
</script>
```

My App has many components,I want to pack them into one file,how to do ?

**Comment by Rich-Harris** [comment]
Essentially, you need to convert all your `.html` files into `.js` files, and then use a module bundler (not included!) to bundle them up into a single file.

Are you already using a build tool in your app? If so, there might be a relatively straightforward way to set this up. If not, the process is slightly cumbersome; bear with me.

First, you can use [rcu](https://github.com/ractivejs/rcu) and [rcu-builders](https://github.com/ractivejs/rcu-builders) (these are tools used by higher-level component loaders) to convert the contents of the `.html` files:

```js
// in a node script somewhere...
var fs = require( 'fs' );
var rcu = require( 'rcu' );
var builders = require( 'rcu-builders' );
var Ractive = require( 'ractive' );

// Initialise ractive component utils
rcu.init( Ractive );

fs.readdirSync( 'components' ).forEach( function ( htmlFile ) {
  var html = fs.readFileSync( 'components/' + htmlFile ).toString();
  var component = rcu.parse( html );

  // now, we create a JavaScript module
  var commonJs = builders.cjs( component );
  fs.writeFileSync( 'compiled-components/' + htmlFile.replace( '.html', '.js' ) );
});
```

Once you've done that, you have a folder (`compiled-components`) full of `.js` files that you can `require()` like any other JavaScript module. From there, you could create an index file like this...

```js
// components.js
module.exports = {
  Foo: require( './compiled-components/Foo' ), // include Foo.html
  Bar: require( './compiled-components/Bar' ), // include Bar.html
  /* and so on */
};
```

...which you can create the bundle from. For example with [Browserify](http://browserify.org/), you could do this...

```bash
browserify components.js -o bundle.js --standalone myComponents
```

...then you could simply include that script on your page:

```html
<script src='bundle.js'></script> <!-- this bundle includes Ractive itself -->
<script>
  var foo = new myComponents.Foo({...});
</script>
```

Going to mark this with the 'enhancement' label, as it'd be nice if there were a command line tool that automated all of this stuff.

**Comment by blueforest** [comment]
According to your idea,I wrote a [demo](https://github.com/blueforest/ractiveComponents),It works well.But I think it's a little more complicated.Are there any other solutions? 
:)

**Comment by fskreuz** [comment]
> My App has many components,I want to pack them into one file,how to do ?

There's also [rvc](https://github.com/ractivejs/rvc). The way I did it in our project was to just write it as components per file. We had close to a thousand components after a few months. This wasn't an issue, since a bug in a component becomes isolated to the component (was a maintenance advantage). [RequireJS](http://requirejs.org/) and rvc did the rest of the heavy lifting in module management.

During release, we run a build prior to deploy. It runs the main file (where the initial `require` call was placed) into the [RequireJS optimizer](http://requirejs.org/docs/optimization.html). Optimizer traces all the dependencies and wraps them up into one file which then replaces the main file.

All of the work was done in JS and the browser. The HTML does not know the main file (with the initial require call) was replaced by a bundled version of all components. There was no HTML replacement magic, nor some build tool that kept watch of file changes.
